parsers/NED.py

```diff
@@ -0,0 +1,224 @@
+
+URL = "https://api.ned.nl/v1/utilizations"
+
+TYPE_MAPPING = {
+    2: "solar",
+}
+
+
+class NedType(Enum):
+    SOLAR = 2
+
+
+class NedActivity(Enum):
+    PRODUCTION = 1
+    CONSUMPTION = 2
+
+
+class NedGranularity(Enum):
+    TEN_MINUTES = 3
+    FIFTEEN_MINUTES = 4
+    HOURLY = 5
+    DAILY = 6
+    MONTHLY = 7
+    YEARLY = 8
+
+
+class NedGranularityTimezone(Enum):
+    UTC = 0
+    LOCAL = 1
+
+
+class NedClassification(Enum):
+    FORECAST = 1
+    MEASURED = 2
+
+
+class NedPoint(Enum):
+    NETHERLANDS = 0
+
+
+# kWh to MWh with 3 decimal places
+def _kwh_to_mw(kwh):
+    return round((kwh / 1000) * 4, 3)
+
+
+# There seems to be a limitation of 144 items we can get in the response in the API at a time
+# So we need to query each mode separately and then combine them
+def call_api(target_datetime: datetime, forecast: bool = False):
+    params = {
+        "itemsPerPage": 192,
+        "point": NedPoint.NETHERLANDS.value,
+        "type[]": NedType.SOLAR.value,
+        "granularity": NedGranularity.FIFTEEN_MINUTES.value,
+        "granularitytimezone": NedGranularityTimezone.UTC.value,
+        "classification": NedClassification.FORECAST.value
+        if forecast
+        else NedClassification.MEASURED.value,
+        "activity": NedActivity.PRODUCTION.value,
+        "validfrom[before]": (target_datetime + timedelta(days=2 if forecast else 1))
+        .date()
+        .isoformat(),
+        "validfrom[after]": (target_datetime - timedelta(days=0 if forecast else 1))
+        .date()
+        .isoformat(),
+    }
+    headers = {"X-AUTH-TOKEN": get_token("NED_KEY"), "accept": "application/json"}
+    response = requests.get(URL, params=params, headers=headers)
+    if not response.ok:
+        raise ParserException(
+            parser="NED.py",
+            message=f"Failed to fetch NED data: {response.status_code}, err: {response.text}",
+        )
+    return response.json()
+
+
+def format_data(json: Any, logger: Logger, forecast: bool = False):
+    df = pd.DataFrame(json)
+    df.drop(
+        columns=[
+            "id",
+            "point",
+            "classification",
+            "activity",
+            "granularity",
+            "granularitytimezone",
+            "emission",
+            "emissionfactor",
+            "capacity",
+            "validto",
+            "lastupdate",
+        ],
+        inplace=True,
+    )
+
+    df = df.groupby(by="validfrom")
+
+    formatted_production_data = ProductionBreakdownList(logger)
+    for _group_key, group_df in df:
+        data_dict = group_df.to_dict(orient="records")
+        mix = ProductionMix()
+        for data in data_dict:
+            clean_type = int(data["type"].split("/")[-1])
+            if clean_type in TYPE_MAPPING:
+                mix.add_value(
+                    TYPE_MAPPING[clean_type],
+                    _kwh_to_mw(data["volume"]),
+                )
+            else:
+                logger.warning(f"Unknown type: {clean_type}")
```

Comment:
This is a bit overkill for just solar but makes it easy to add other modes as well in the future, like wind for example.